File: octo/msgp/msgp-server/src/main/scala/com/sankuai/octo/msgp/utils/client/MnsCacheClient.scala
package com.sankuai.octo.msgp.utils.client

import com.sankuai.octo.mnsc.idl.thrift.service.MNSCacheService
import org.springframework.beans.BeansException
import org.springframework.beans.factory.{FactoryBean, InitializingBean}
import org.springframework.context.{ApplicationContext, ApplicationContextAware}

object MnsCacheClient {
  private var client: MNSCacheService.Iface = null

  def getInstance: MNSCacheService.Iface = {
    client
  }
}

class MnsCacheClient extends FactoryBean[AnyRef] with ApplicationContextAware with InitializingBean {
  private var clientProxy: AnyRef = null

  @throws(classOf[BeansException])
  def setApplicationContext(applicationContext: ApplicationContext) {
  }

  @throws(classOf[Exception])
  def getObject: AnyRef = {
    this
  }

  def getObjectType: Class[_] = {
    this.getClass
  }

  def isSingleton: Boolean = {
    true
  }

  @throws(classOf[Exception])
  def afterPropertiesSet {
    initInstance
  }

  def getClientProxy: AnyRef = {
    clientProxy
  }

  def setClientProxy(clientProxy: AnyRef) {
    this.clientProxy = clientProxy
  }

  private def initInstance {
    MnsCacheClient.client = clientProxy.asInstanceOf[MNSCacheService.Iface]
  }
}

File: octo/mns/scanner/src/main/java/com/sankuai/octo/scanner/service/MNSCacheClient.java
package com.sankuai.octo.scanner.service;

import com.sankuai.octo.mnsc.idl.thrift.service.MNSCacheService;
import org.springframework.beans.BeansException;
import org.springframework.beans.factory.FactoryBean;
import org.springframework.beans.factory.InitializingBean;
import org.springframework.context.ApplicationContext;
import org.springframework.context.ApplicationContextAware;


/**
 * Copyright (C) 2015 Meituan
 * All rights reserved
 * User: gaosheng
 * Date: 15-9-22
 * Time: 下午5:55
 */
public class MNSCacheClient implements FactoryBean<Object>, ApplicationContextAware, InitializingBean {

    private Object clientProxy;
    private static MNSCacheService.Iface client;


    @Override
    public void setApplicationContext(
            ApplicationContext applicationContext) throws BeansException {
    }

    @Override
    public Object getObject() throws Exception {
        return this;
    }

    @Override
    public Class<?> getObjectType() {
        return this.getClass();
    }

    @Override
    public boolean isSingleton() {
        return true;
    }

    @Override
    public void afterPropertiesSet() throws Exception {
        initInstance();
    }

    public Object getClientProxy() {
        return clientProxy;
    }

    public void setClientProxy(Object clientProxy) {
        this.clientProxy = clientProxy;
    }

    private void initInstance() {
        client = (MNSCacheService.Iface) clientProxy;
    }

    public static MNSCacheService.Iface getInstance() {
        return client;
    }
}

